File: 0367_valid_perfect_square/is_perfect_square.go
package main

/**
 * Time: O(log(n))
 * Space: O(1)
 */
func isPerfectSquare(num int) bool {
    if num < 2 {
        return true
    }
    l, r := 2, num / 2
    for l <= r {
        m := (l + r) / 2
        squared := m * m
        if squared == num {
            return true
        }
        if squared > num {
            r = m - 1
        } else if squared < num {
            l = m + 1
        }
    }
    return false
}

func main() {

}

File: 0367_valid_perfect_square/is_perfect_square.py
def is_perfect_square(num: int) -> bool:
    """
    Time: O(log(n))
    Space: O(1)
    """
    if num < 2:
        return True
    left, right = 2, num // 2
    while left <= right:
        mid = (left + right) // 2
        squared = mid * mid
        if squared == num:
            return True
        if squared > num:
            right = mid - 1
        elif squared < num:
            left = mid + 1
    return False
